### DISCUSSION TITLE
Energy generation.

### DISCUSSION THREAD

**Discussion by JackCrumpLeys** [DigitalExtinction/Game#466] [Ideas]
Here is rough overview of all of the tasks I believe need to be completed to have a good base of an energy system. I have annotated it with '--' where I want some input/further discussion. Over the coming months I will try to work to complete this goal.

1. Energy Generation
	- a. Design and finalise the specifications for each type of power plant (solar, nuclear, wind).
		- Research real-world power plant efficiency, environmental factors, and limitations that are applicable in-game. -- What would be plausible in this future world?
		- Calculate and document the energy generation rates and attributes for each type of power plant. -- How tied to the real world energy are we going to make it? Are we literally going to use watts or just some arbitrary 'Energy unit'?
		- Define the cost of constructing and maintaining each type of power plant in terms of energy and in-game resources/money.
	- b. Implement power plant models (with 3D assets). This can come later (use placeholder art)
		- Collaborate with artists to create detailed and appealing 3D assets for each type of power plant.
		- Implement the visual and animation components of the power plants using the game engine. --  I anticipate that animation is a whole side-quest in itself.
		- Ensure that the power plant models and animation seamlessly integrate with the game's overall visual style.
	- c. Develop a system to monitor and display energy generation rates for each type.
		- Design and implement a UI element to show energy generation rates on-screen.
		- Integrate the UI element with the energy generation system, ensuring real-time updates.
	- d. Implement the integration of generated energy from power plants into the energy grid.
		- Define how the generated energy flows into the energy distribution system.
		- Balance the input of energy while considering the grid's current load and potential expansion.

2. Energy Distribution
	- a. Design a laser-based grid system for energy distribution across buildings and units.
		- Determine the properties and constraints of the laser-based grid system, such as the range and maximum power transfer. -- Is energy transfer lossless?
		- Outline the possible setups and configurations of the grid system that will allow for versatile power distribution in different game scenarios.
		- Create the visual design, such as colours, effects, and animation, to indicate an active and functional energy distribution grid. The grid should be brighter the more power it is transferring. Possible toggle-able UI to show the amount of power in a laser connection.
	
	- b. Develop algorithms to handle the maximum energy flow optimisation problem.
		- Research and analyse algorithms that are suitable for handling energy distribution in large-scale grids. [https://en.wikipedia.org/wiki/Maximum_flow_problem](https://en.wikipedia.org/wiki/Maximum_flow_problem "https://en.wikipedia.org/wiki/Maximum_flow_problem")
		- Implement chosen algorithm to accommodate the game's specific requirements, such as the energy generation, storage, and consumption rates.
		- Test the algorithm's performance under various conditions and iterate as needed to ensure efficiency and responsiveness during game-play.
	
	- c. Implement power hubs as the primary energy distribution method.
		- Design the specifications and attributes of power hubs, such as the number of energy lasers and energy panels, and battery requirements. -- these may be up-gradable by research or construction.
		- Integrate the power hubs into the laser-based grid system, enabling them to transmit and receive energy from connected buildings and units.
		- Provide a method for players to construct, upgrade, and remove power hubs from the grid system, allowing for dynamic changes during game-play.
	
	- d. Implement energy lasers and energy panels for buildings and units, enabling them to transmit and receive energy.
		- Determine the properties of energy lasers and energy panels, such as energy transfer rates, range limits, and line of sight requirements. -- Are Enemy units able to hijack them by moving in front of them? Are we going to have unit(s) that are interceptors and act as a mirror to reflect energy into their own grid?
		- Create 3D assets for energy lasers and energy panels. -- automatic addition of receiver asset on models (this way we may be able to have different levels of receivers on units without making a bunch of models for units)?
		- Implement the behaviour and mechanics of energy lasers and energy panels, allowing them to establish and maintain connections with power hubs and other targets.
	
	- e. Design and implement idle-state rules for grid management (upgrading, repairing, and cleaning).
		- Define how the energy grid behaves during idle states (e.g., when there is an excess or deficiency of energy).
		- Implement systems to automatically upgrade, repair, or re-purpose buildings and units to optimise the energy grid during idle states. This should be fully configurable at a macro and a micro level and have sensible defaults. -- I have a lot of ideas surrounding this system and am keen to discuss them but should probably cross the bridge when we get to it.

3. Energy Storage
	- a. Design battery farms and integrated batteries systems for buildings and units, including max capacity and charge rates.
		- Research real-world battery technology to gain insights into capacity, charging rates, and discharge properties.
		- Determine the design considerations, such as size, weight, and other in-game factors, for each type of battery.
		- Calculate the specific attributes, such as max capacity and charge rates, for use by various buildings and units.
	- b. Model and implement battery farms and integrate them into the energy grid.
		- Collaborate with artists to create appealing 3D assets for the battery farms and integrated batteries. This can come later (use placeholder art).
		- Implement the visual and animation components of the battery models in the game engine.
		- Set up the communication between battery farms and the energy distribution system for proper energy storage and usage.
	- c. Implement a system for displaying energy storage levels and charging rates for all batteries in the grid.
		- Create user-friendly UI elements to display battery information with real-time updates.
		- Customize the UI elements for each type of energy storage system to distinguish them visually and functionally.
	- d. Create rules for managing stored energy, including prioritization, discharge rates, and emergencies.
		- Define how the energy storage system prioritizes energy distribution during different scenarios.
		- Calculate the optimal discharge rates to ensure a balanced and efficient energy supply.
		- Implement emergency rules that dictate the distribution of energy in times of crisis, such as during attacks or resource shortages. For example if under attack maybe repair priority increases. -- Should we make a system that automatically moves power from batteries that are under threat?

4. Polish and Testing
	- a. Perform thorough testing on all aspects of the energy system to ensure balance and optimal performance.
		- Test all elements of the energy system in isolation to ensure proper and efficient functioning.
		- Conduct integrated tests to ensure that all elements of the energy system work together harmoniously.
		- Iterate on the energy system's balance, making adjustments as needed to create a fair and engaging game-play experience.
	- b. Refactor and optimise the code, as necessary.
		- Review the code for optimisation opportunities and implement changes to improve performance.
		- Ensure that the code adheres to established development guidelines and best practices.
	- c. Document all implemented systems and create user tutorials for understanding the game mechanics.
		- Write detailed documentation for the development team and future contributors to understand the energy system components.
		- Develop tutorials and in-game help systems to make the energy system accessible to players.

**Comment by Indy2222** [comment]
Thank you for the elaboration on the topic ❤️.

I think we should start very, very small. Let's implement each little piece, test it and decide on next step. It will be a huge amount of work.

In the very first version, it is enough to:
 
1. Attach a battery to every unit and building. The battery continually (dis)charges at a rate which is a sum of a) constant discharge rate (independent on anything), b) charging rate from a connected power hub.
1. The unit / building stops working when the battery dies. It starts working again once it reaches some charge.
1. Player can construct power hubs. They form a graph. Any two power hubs or a unit/building and a power hub with distance less than X are connected.
1. Each connection / interlink has a maximum power (in watts). This is a constant.
1. Player can construct a "power plant". It continually produces Y watts of power.
1. Power distribution is limited by the distribution graph (i.e. interlinks and their maximum power). If the grid's capacity goes beyond power plant's power, it is distributed "at random" (whatever is easiest). There is no optimization beyond maximum flow.

> Here is rough overview of all of the tasks I believe need to be completed to have a good base of an energy system. I have annotated it with '--' where I want some input/further discussion. Over the coming months I will try to work to complete this goal.
> 
>     1. Energy Generation
>        
>        * a. Design and finalise the specifications for each type of power plant (solar, nuclear, wind).

While I agree that we should have a better "vision" before the implementation starts, I think it is not necessary to nail this down right away. I would start with something very simple, e.g. a solar power plant, do the bare bones design & implementation and then build from there incrementally.

>          
>          * Research real-world power plant efficiency, environmental factors, and limitations that are applicable in-game. -- What would be plausible in this future world?

A table / an article with all current and "near-future" electricity generation technologies would be nice. Something basic would be sufficient. For example median construction time, construction cost, generation cost (per kWh), power output. We can use this to decide what to include in the game.

>          * Calculate and document the energy generation rates and attributes for each type of power plant. -- How tied to the real world energy are we going to make it? Are we literally going to use watts or just some arbitrary 'Energy unit'?

I would use Watts (for power) and Joules (for energy). However, we might need to scale power/energy (or pace of time) so that it works well in the game environment. Similarly to how construction cannot be real-time for obvious reasons.

>          * Define the cost of constructing and maintaining each type of power plant in terms of energy and in-game resources/money.

I would like to build the game mostly around energy. Therefore, there will be no other resources in the initial versions. Once energy is implemented, fine tuned and well tested for playability we can decide on next steps (for example introducing general "material" which can be mined).

>        * b. Implement power plant models (with 3D assets). This can come later (use placeholder art)
>          
>          * Collaborate with artists to create detailed and appealing 3D assets for each type of power plant.
>          * Implement the visual and animation components of the power plants using the game engine. --  I anticipate that animation is a whole side-quest in itself.

Yes. I would not couple these tasks.

>          * Ensure that the power plant models and animation seamlessly integrate with the game's overall visual style.
>        * c. Develop a system to monitor and display energy generation rates for each type.
>          
>          * Design and implement a UI element to show energy generation rates on-screen.
>          * Integrate the UI element with the energy generation system, ensuring real-time updates.

Again, I would start with something extremely bare bones. Step by step.

>        * d. Implement the integration of generated energy from power plants into the energy grid.
>          
>          * Define how the generated energy flows into the energy distribution system.
>          * Balance the input of energy while considering the grid's current load and potential expansion.
> 
>     2. Energy Distribution
>        
>        * a. Design a laser-based grid system for energy distribution across buildings and units.
>          
>          * Determine the properties and constraints of the laser-based grid system, such as the range and maximum power transfer. -- Is energy transfer lossless?

I would start with lossless. The only limitations (at this point) should be maximum power per link and maximum distance. We can later open discussion on other more complex topics, e.g. line of sight requirements, laser vs induction, efficiency, etc.

>          * Outline the possible setups and configurations of the grid system that will allow for versatile power distribution in different game scenarios.
>          * Create the visual design, such as colours, effects, and animation, to indicate an active and functional energy distribution grid. The grid should be brighter the more power it is transferring. Possible toggle-able UI to show the amount of power in a laser connection.
>        * b. Develop algorithms to handle the maximum energy flow optimisation problem.
>          
>          * Research and analyse algorithms that are suitable for handling energy distribution in large-scale grids. https://en.wikipedia.org/wiki/Maximum_flow_problem
>          * Implement chosen algorithm to accommodate the game's specific requirements, such as the energy generation, storage, and consumption rates.

We should put a lot of energy into this so it goes well with Bevy ECS. I think that Bevy's entity hierarchy can be an inspiration here.

>          * Test the algorithm's performance under various conditions and iterate as needed to ensure efficiency and responsiveness during game-play.
>        * c. Implement power hubs as the primary energy distribution method.
>          
>          * Design the specifications and attributes of power hubs, such as the number of energy lasers and energy panels, and battery requirements. -- these may be up-gradable by research or construction.

Yes. I would make everything fixed (e.g. the number of lasers) at this point and decide on next steps later. In-game R&D is an independent and large topic by itself so lets not couple these.

>          * Integrate the power hubs into the laser-based grid system, enabling them to transmit and receive energy from connected buildings and units.
>          * Provide a method for players to construct, upgrade, and remove power hubs from the grid system, allowing for dynamic changes during game-play.

What about allowing only construction and leaving grid utilization optimization on the "maximum flow" algorithm?

>        * d. Implement energy lasers and energy panels for buildings and units, enabling them to transmit and receive energy.
>          
>          * Determine the properties of energy lasers and energy panels, such as energy transfer rates, range limits, and line of sight requirements. -- Are Enemy units able to hijack them by moving in front of them? Are we going to have unit(s) that are interceptors and act as a mirror to reflect energy into their own grid?

I would love these features but not in the first version.

>          * Create 3D assets for energy lasers and energy panels. -- automatic addition of receiver asset on models (this way we may be able to have different levels of receivers on units without making a bunch of models for units)?

Ditto, good idea for a future work.

>          * Implement the behaviour and mechanics of energy lasers and energy panels, allowing them to establish and maintain connections with power hubs and other targets.
>        * e. Design and implement idle-state rules for grid management (upgrading, repairing, and cleaning).
>          
>          * Define how the energy grid behaves during idle states (e.g., when there is an excess or deficiency of energy).
>          * Implement systems to automatically upgrade, repair, or re-purpose buildings and units to optimise the energy grid during idle states. This should be fully configurable at a macro and a micro level and have sensible defaults. -- I have a lot of ideas surrounding this system and am keen to discuss them but should probably cross the bridge when we get to it.

Ditto, IMHO too complex for an early version.

>     3. Energy Storage
>        
>        * a. Design battery farms and integrated batteries systems for buildings and units, including max capacity and charge rates.
>          
>          * Research real-world battery technology to gain insights into capacity, charging rates, and discharge properties.
>          * Determine the design considerations, such as size, weight, and other in-game factors, for each type of battery.
>          * Calculate the specific attributes, such as max capacity and charge rates, for use by various buildings and units.
>        * b. Model and implement battery farms and integrate them into the energy grid.
>          
>          * Collaborate with artists to create appealing 3D assets for the battery farms and integrated batteries. This can come later (use placeholder art).
>          * Implement the visual and animation components of the battery models in the game engine.
>          * Set up the communication between battery farms and the energy distribution system for proper energy storage and usage.
>        * c. Implement a system for displaying energy storage levels and charging rates for all batteries in the grid.
>          
>          * Create user-friendly UI elements to display battery information with real-time updates.
>          * Customize the UI elements for each type of energy storage system to distinguish them visually and functionally.
>        * d. Create rules for managing stored energy, including prioritization, discharge rates, and emergencies.
>          
>          * Define how the energy storage system prioritizes energy distribution during different scenarios.
>          * Calculate the optimal discharge rates to ensure a balanced and efficient energy supply.
>          * Implement emergency rules that dictate the distribution of energy in times of crisis, such as during attacks or resource shortages. For example if under attack maybe repair priority increases. -- Should we make a system that automatically moves power from batteries that are under threat?
> 
>     4. Polish and Testing
>        
>        * a. Perform thorough testing on all aspects of the energy system to ensure balance and optimal performance.
>          
>          * Test all elements of the energy system in isolation to ensure proper and efficient functioning.
>          * Conduct integrated tests to ensure that all elements of the energy system work together harmoniously.
>          * Iterate on the energy system's balance, making adjustments as needed to create a fair and engaging game-play experience.
>        * b. Refactor and optimise the code, as necessary.
>          
>          * Review the code for optimisation opportunities and implement changes to improve performance.
>          * Ensure that the code adheres to established development guidelines and best practices.
>        * c. Document all implemented systems and create user tutorials for understanding the game mechanics.
>          
>          * Write detailed documentation for the development team and future contributors to understand the energy system components.
>          * Develop tutorials and in-game help systems to make the energy system accessible to players.

**Comment by Indy2222** [comment]
This might come handy when/if it lands https://github.com/bevyengine/bevy/issues/3742

**Comment by Indy2222** [comment]
Actual implementation of the first version of energy is tracked by #256.

**Comment by JackCrumpLeys** [comment]
🎉 basic battery functionality is merged #483!
